src/Data/Array/Accelerate/AST.hs

```diff
@@ -559,6 +561,21 @@ data OpenExp env aenv t where
                 -> OpenExp env aenv (Vec n s)
                 -> OpenExp env aenv tup
 
+  VecIndex      :: (KnownNat n, v ~ Vec n s)
+                => VectorType v
+                -> IntegralType i
+                -> OpenExp env aenv (Vec n s)
+                -> OpenExp env aenv i
+                -> OpenExp env aenv s
+
+  VecWrite      :: (KnownNat n, v ~ Vec n s)
+                => VectorType v
+                -> IntegralType i
+                -> OpenExp env aenv (Vec n s)
+                -> OpenExp env aenv i
+                -> OpenExp env aenv s
+                -> OpenExp env aenv (Vec n s)
+
```

Comment:
Remove `VecPac` and `VecUnpack`? Those just get turned your new index and write instructions. We might as well also add `VecShuffle`, corresponding to the [`shufflevector`](https://llvm.org/docs/LangRef.html#shufflevector-instruction) instruction.

Reply:
One of the more complicated things about VecPack and Unpack is the need to derive the tuple type. I got rather stuck trying to make changes to that whole pipeline, could you assist me on this live sometime?

Reply:
Ah, good point. Okay let's just shelve that for now.

I've been thinking how this vector support can be restructured after you mentioned you couldn't have a `Vec Node` (or any non-primitive thing), which is sort of a weird limitation for our language...

Reply:
The two blocking constraints are `Prim Node` (which can be implemented I think. However `EltR Node ~ Node` is the culprit. Perhaps some unsafe coerceon from and to a flattened byte vector?